src/librustc/mir/interpret/mod.rs

```diff
@@ -348,42 +350,60 @@ impl<'tcx, M: fmt::Debug + Eq + Hash + Clone> AllocMap<'tcx, M> {
         id
     }
 
-    // FIXME: Check if functions have identity. If not, we should not intern these,
-    // but instead create a new id per use.
-    // Alternatively we could just make comparing function pointers an error.
+    /// Functions cannot be identified by pointers, as asm-equal functions can get deduplicated
+    /// by the linker and functions can be duplicated across crates.
+    /// We thus generate a new `AllocId` for every mention of a function. This means that
+    /// `main as fn() == main as fn()` is false, while `let x = main as fn(); x == x` is true.
     pub fn create_fn_alloc(&mut self, instance: Instance<'tcx>) -> AllocId {
-        self.intern(AllocType::Function(instance))
+        let id = self.reserve();
+        self.id_to_type.insert(id, AllocKind::Function(instance));
+        id
     }
 
-    pub fn get(&self, id: AllocId) -> Option<AllocType<'tcx, M>> {
+    /// Returns `None` in case the `AllocId` is dangling.
```

Comment:
"Dangling" for the *global* memory only, right? It could still be allocated in the local (`EvalContext`) memory?

Reply:
Yes, but the local memory will never hit the global one if it has that `AllocId` locally.

Reply:
Sure, I just think the comment should explain what "dangling" means here.